external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/FetchedDataPool.scala

```diff
@@ -0,0 +1,160 @@
+    var inUse: Boolean = false
+
+    def getObject: FetchedData = fetchedData
+  }
+
+  private object CachedFetchedData {
+    def empty(): CachedFetchedData = {
+      val emptyData = FetchedData(
+        ju.Collections.emptyListIterator[ConsumerRecord[Array[Byte], Array[Byte]]],
+        UNKNOWN_OFFSET,
+        UNKNOWN_OFFSET)
+
+      CachedFetchedData(emptyData)
+    }
+  }
+
+  private type CachedFetchedDataList = mutable.ListBuffer[CachedFetchedData]
+
+  private val cache: mutable.Map[CacheKey, CachedFetchedDataList] = mutable.HashMap.empty
+
+  /** Retrieve internal cache. This method is only for testing. */
+  private[kafka010] def getCache: mutable.Map[CacheKey, CachedFetchedDataList] = cache
+
+  private val (minEvictableIdleTimeMillis, evictorThreadRunIntervalMillis): (Long, Long) = {
+    val conf = SparkEnv.get.conf
+
+    val minEvictIdleTime = conf.getLong(CONFIG_NAME_MIN_EVICTABLE_IDLE_TIME_MILLIS,
+      DEFAULT_VALUE_MIN_EVICTABLE_IDLE_TIME_MILLIS)
+
+    val evictorThreadInterval = conf.getLong(
+      CONFIG_NAME_EVICTOR_THREAD_RUN_INTERVAL_MILLIS,
+      DEFAULT_VALUE_EVICTOR_THREAD_RUN_INTERVAL_MILLIS)
+
+    (minEvictIdleTime, evictorThreadInterval)
+  }
+
+  private val executorService = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+    "kafka-fetched-data--cache-evictor")
+
+  private def startEvictorThread(): Unit = {
+    executorService.scheduleAtFixedRate(new Runnable() {
+      override def run(): Unit = {
+        removeIdleFetchedData()
+      }
+    }, 0, evictorThreadRunIntervalMillis, TimeUnit.MILLISECONDS)
+  }
+
+  startEvictorThread()
+
+  def acquire(key: CacheKey, desiredStartOffset: Long): FetchedData = synchronized {
+    val fetchedDataList = cache.getOrElseUpdate(key, new CachedFetchedDataList())
+
+    val cachedFetchedDataOption = fetchedDataList.find { p =>
+      !p.inUse && p.getObject.nextOffsetInFetchedData == desiredStartOffset
+    }
+
+    var cachedFetchedData: CachedFetchedData = null
+    if (cachedFetchedDataOption.isDefined) {
+      cachedFetchedData = cachedFetchedDataOption.get
+    } else {
+      cachedFetchedData = CachedFetchedData.empty()
+      fetchedDataList += cachedFetchedData
+    }
+
+    cachedFetchedData.lastAcquiredTimestamp = System.currentTimeMillis()
+    cachedFetchedData.inUse = true
+
+    cachedFetchedData.getObject
+  }
+
+  def invalidate(key: CacheKey): Unit = synchronized {
+    cache.remove(key)
+  }
+
+  def release(key: CacheKey, fetchedData: FetchedData): Unit = synchronized {
+    cache.get(key) match {
+      case Some(fetchedDataList) =>
+        val cachedFetchedDataOption = fetchedDataList.find { p =>
+          p.inUse && p.getObject == fetchedData
+        }
+
+        if (cachedFetchedDataOption.isDefined) {
+          val cachedFetchedData = cachedFetchedDataOption.get
+          cachedFetchedData.inUse = false
+          cachedFetchedData.lastReleasedTimestamp = System.currentTimeMillis()
+        }
+
+      case None =>
```

Comment:
Yeah sometimes yes and sometimes ignorable thing, I intended to avoid the situation which end users are confused with false-negative, but I'm also open to add log message for that.

Let me add some warning message, and if some reviewer stated the level or removing log, we can revisit it later.

Reply:
Thanks, yeah just for the record I meant something like this: https://github.com/apache/spark/blob/341b55a58964b1966a1919ac0774c8be5d5e7251/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala#L623